### PR TITLE
Changes from feedback on gRFC L82: gRPC Core EventEngine API

### PR DESCRIPTION
**Pull request by drfloob**
Related discussions are in https://github.com/grpc/proposal/pull/245

The SliceAllocator and SliceAllocatorFactory interface changes are required for https://github.com/grpc/grpc/pull/26643.

CC @ctiller @nicolasnoble

### CODE REVIEW

File: src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.cc [hunk 1]
```diff
@@ -17,71 +17,74 @@
...
 
 namespace grpc_event_engine {
 namespace experimental {
 
-SliceAllocator::SliceAllocator(grpc_resource_user* user)
-    : resource_user_(user) {
-  grpc_resource_user_ref(resource_user_);
-};
+Chttp2SliceAllocator::Chttp2SliceAllocator(grpc_resource_user* user)
+    : resource_user_(user){};
```
Comment (markdroth) [inline src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.cc:30]: Nit: No need for the semicolon here.

File: src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.cc [hunk 2]
```diff
@@ -17,71 +17,74 @@
...
 
-SliceAllocator::~SliceAllocator() {
+Chttp2SliceAllocator::~Chttp2SliceAllocator() {
   if (resource_user_ != nullptr) {
     grpc_resource_user_unref(resource_user_);
   }
 };
```
Comment (markdroth) [inline src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.cc:36]: No need for semicolon here either.

File: src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+class Chttp2SliceAllocator
+    : public grpc_event_engine::experimental::SliceAllocator {
+ public:
+  /// gRPC-internal constructor. Takes ownership of a resource_user ref from the
+  /// caller.
+  explicit Chttp2SliceAllocator(grpc_resource_user* user);
+  // Not copyable
+  Chttp2SliceAllocator(Chttp2SliceAllocator& other) = delete;
+  Chttp2SliceAllocator& operator=(const Chttp2SliceAllocator& other) = delete;
+  // Moveable
```
Comment (markdroth) [inline src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.h:39]: Now that it's going to be returned as a `unique_ptr<>`, it doesn't actually need to be movable anymore.

File: src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+class Chttp2SliceAllocatorFactory
+    : public grpc_event_engine::experimental::SliceAllocatorFactory {
+ public:
+  // gRPC-internal constructor
+  explicit Chttp2SliceAllocatorFactory(grpc_resource_quota* quota);
+  // Not copyable
+  Chttp2SliceAllocatorFactory(Chttp2SliceAllocatorFactory& other) = delete;
+  Chttp2SliceAllocatorFactory& operator=(
+      const Chttp2SliceAllocatorFactory& other) = delete;
+  // Moveable
```
Comment (markdroth) [inline src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.h:59]: Same here: doesn't actually need to be movable.

File: src/core/lib/iomgr/event_engine/tcp.cc [hunk 5]
```diff
@@ -38,28 +38,43 @@ using ::grpc_event_engine::experimental::EventEngine;
...
+  std::unique_ptr<SliceAllocator> CreateSliceAllocator(
+      absl::string_view peer_name) {
+    return absl::make_unique<NoopSliceAllocator>();
+  };
+};
+
 struct grpc_tcp_server {
-  grpc_tcp_server(std::unique_ptr<EventEngine::Listener> listener,
-                  grpc_resource_quota* rq)
+  grpc_tcp_server(std::unique_ptr<EventEngine::Listener> listener)
```
Comment (markdroth) [inline src/core/lib/iomgr/event_engine/tcp.cc:64]: This should be `explicit`, since there's now only one argument.

**Review by markdroth — APPROVED** [review]
This looks good!  My comments are mostly cosmetic.

Please let me know if you have any questions.  Thanks!

File: src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.h (hunk 3 above)
Comment (drfloob) [reply src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.h]: Done

File: src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.h (hunk 4 above)
Comment (drfloob) [reply src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.h]: Done

**Review by drfloob — COMMENTED** [review]
Thanks! I've made the changes, and added another from the gRFC: having Endpoint GetPeerAddress and GetLocalAddress return `ResolvedAddress&` instead of pointers per the style guide.

File: src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.cc (hunk 2 above)
Comment (drfloob) [reply src/core/ext/transport/chttp2/transport/chttp2_slice_allocator.cc]: Done

**Review by markdroth — APPROVED** [review]
Looks great!

**Comment by drfloob** [comment]
interop cloud to prod: #22683

**Pull request merged**